Added functionality to handle unseen levels in MeanResponseTransformer

Ritunjai-Sharma opened the pull request:
Added functionality to handle unseen levels in MeanResponseTransformer. User can choose on of the following unseen_levels for imputing unseen levels : 'Mean', 'Median', 'Lowest', 'Highest' or can provide an arbitrary int/float value to impute all unseen levels. The default value of 'None' will output error if the data to use transform on contains unseen levels.

davidhopkinson26 commented:
Thanks for the contribution Ritunjai! There are a few things to add before this can be merged in:

1. add testing of new functionality
2. update the changelog with the change (add in a 0.3.5 unreleased section please as explained in the changelog doc)
3. add an example to the example notebook for this transformer

davidhopkinson26 commented on tubular/nominal.py:
```diff
@@ -537,11 +546,20 @@ class MeanResponseTransformer(BaseNominalTransformer, BaseMappingTransformMixin)
         Only created in the mutli-level case. A dictionary of the form level : transformer containing the mean response
         transformers for each level to be encoded against.
 
+    impute_dict: dict
```
I think we should avoid using 'impute' as this is typically used in reference to null values and may confuse a user.  `unseen_levels_encoding_dict` might be clearer?

Ritunjai-Sharma replied:
Absolutely agree! Replaced impute_dict with unseen_levels_encoding_dict in the commit: [3307aa84](https://github.com/lvgig/tubular/pull/80/commits/3307aa84031d6356f7e42f62d4bb1e41f70f949d)

davidhopkinson26 commented on tubular/nominal.py:
```diff
@@ -537,11 +546,20 @@ class MeanResponseTransformer(BaseNominalTransformer, BaseMappingTransformMixin)
+        weights_column=None,
+        prior=0,
+        level=None,
+        unseen_levels=None,
```
`unseen_level_handling` would be more pythonic and make some later code more readable

Ritunjai-Sharma replied:
Yes! unseen_level_handling is much better! Changed unseen_levels to unseen_level_handling in commit: [3307aa84](https://github.com/lvgig/tubular/pull/80/commits/3307aa84031d6356f7e42f62d4bb1e41f70f949d)

Ritunjai-Sharma commented on tubular/nominal.py:
```diff
@@ -502,6 +504,13 @@ class MeanResponseTransformer(BaseNominalTransformer, BaseMappingTransformMixin)
+        data when using transform method. Default value of None will output error when attempting to use transform
+        on data with unseen levels in categorical columns to encode. Set this parameter to one of the options above
+        in order to impute the unseen levels in each categorical column to encode, according to mean, median etc. of
+        each encoded column. One can also pass an arbitrary int/float value to use for imputing all unseen levels.
```
Yup, makes sense! Updated the docstring in commit: [3307aa84](https://github.com/lvgig/tubular/pull/80/commits/3307aa84031d6356f7e42f62d4bb1e41f70f949d)

Ritunjai-Sharma commented on tubular/nominal.py:
```diff
@@ -739,10 +765,32 @@ def fit(self, X, y):
+            for c in self.encoded_feature_columns:
+                self.impute_dict[c] = X_temp[c].map(self.mappings[c]).mean()
+
+        if self.unseen_levels == "Median":
```
Thanks for pointing out this error. Changed if to elif in commit: [3307aa84](https://github.com/lvgig/tubular/pull/80/commits/3307aa84031d6356f7e42f62d4bb1e41f70f949d)

Ritunjai-Sharma commented on tubular/nominal.py:
```diff
@@ -777,8 +824,20 @@ def transform(self, X):
 
-        self.check_mappable_rows(X)
-        X = BaseMappingTransformMixin.transform(self, X)
+        if self.unseen_levels in ["Mean", "Median", "Lowest", "Highest"] or isinstance(
```
Yup! This is much cleaner! Fixed in commit: [3307aa84](https://github.com/lvgig/tubular/pull/80/commits/3307aa84031d6356f7e42f62d4bb1e41f70f949d)

Ritunjai-Sharma commented:
I have updated changelog to include 0.3.5 unreleased section in commit [7993cbc](https://github.com/lvgig/tubular/pull/80/commits/7993cbc332473af3f5f58cb7da639ea3a5759055)

davidhopkinson26 commented:
a note on tests - We are looking to remove tests on implementation and just focus on testing functionality.  Current testing of implementation can inflate coverage.  Please delete the following rather than updating:

test_arguments
test_inheritance
test_class_methods
test_xxxx_called (with exception of test_super_init_called as this covers some error handling)

Ritunjai-Sharma commented:
Thanks for the note on tests. Will delete the required tests.

Ritunjai-Sharma commented:
I have added code for testing unseen level handling in the commit: [d28550e](https://github.com/lvgig/tubular/pull/80/commits/d28550ed89a3e397f890a12bc74cf3549574616f)
I have also added an example of how to use the unseen level handling functionality in the example notebook in the commit: [4be9961](https://github.com/lvgig/tubular/pull/80/commits/4be99619ca8a65db68026516f2f77f79311813c2)

davidhopkinson26 commented on tests/nominal/test_MeanResponseTransformer.py:
```diff
@@ -169,59 +160,22 @@ def test_output2(self):
 
 
 class TestFit:
-    def test_super_fit_called(self, mocker):
```
please add this one back in too (again, implicit testing of error handling done in the parent class method)

davidhopkinson26 commented on tests/nominal/test_MeanResponseTransformer.py:
```diff
@@ -645,45 +609,210 @@ def expected_df_3():
 
         return df
 
-    def test_check_is_fitted_called(self, mocker):
```
again, add this one back in please.

davidhopkinson26 commented on tests/nominal/test_MeanResponseTransformer.py:
```diff
@@ -645,45 +609,210 @@ def expected_df_3():
-            x.transform(df)
+        df["c"] = df["c"].astype("category")
 
-    def test_super_transform_called(self, mocker):
```
and this one

davidhopkinson26 commented:
Please add in some testing that self.unseen_levels_encoding_dict is correctly populated in fit.  Currently only testing as part of the full fit transform cycle outputs.

davidhopkinson26 commented:
please update the version in version.py to 0.3.5

Ritunjai-Sharma commented:
Hi! I have re-added the required tests, added testing to ensure that unseen_levels_encoding_dict is correctly populated in fit and updated version.py to 0.3.5 in commit  [e614207](https://github.com/lvgig/tubular/pull/80/commits/e6142070de6960032992be812facbf3069f9c273)
